Add support for custom rules

taoufik07 opened the pull request.

taoufik07 committed:
Add support for custom rules

Shr3ps committed:
Update README and lint code

Shr3ps reviewed (approved):
Seems god to me,
I have pushed some lint/minor fixes

BzSpi commented on variables.tf:
```diff
@@ -145,3 +145,9 @@ variable "load_balancer_rules_enabled" {
+
+variable "additional_rules" {
+  description = "Additional network security group rules to add. For arguements please refer to https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/network_security_rule#argument-reference"
+  type        = list(any)
```
Use a fully-typed object with `optional` attributes.

taoufik07 replied:
I don't think it's a good idea as `terraform` 1.3 is still in alpha and the `optional` feature is only experimental in 1.2 and it's not advised to enable it in prod with 1.2.

BzSpi replied:
Indeed, that's a good concern !

We've discussed it internally a while ago and we've made the choice to allow the use of the experimental feature in our modules for versions < 1.3.
Also, we will release major versions of our modules for Terraform 1.3+ in case they use the optional implementation.

taoufik07 replied:
I've used the experimental `optional` 🤞, although I wanted to do it in a separate PR after the official release.

As I've used `default` with `optional`, I've updated the required terraform version to 1.3.

Shr3ps replied:
Since Terraform 1.3 is not released yet, we will have to go with 1.2 for now.

I will probably make the changes into your branch directly

taoufik07 replied:
This will not work in 1.2 as I've used the `default` param with `optional`. You can check my commit message for more details.

Shr3ps replied:
Changes pushed

taoufik07 committed:
Revamp and use new `optional` attribute for `additional_rules`

The `optional` attribute is officially available starting from
terraform 1.3.0, which's not yet released.

If `terraform 1.3.0` is still not realeased at the moment you're using
this module, you could use one of the `beta/rc` releases of
`terraform 1.3.0` (`tfenv` is your friend).

Since I'm using `default` with `optional`, this will not work with
`terraform 1.2.8`, which's the latest stable version ATM, even if you
enable the experimental `module_variable_optional_attrs`. Maybe you
can wait for a future `1.2.9`, if there's any, but I'm not sure if
they are planning to include it in the experiment.

I pushed this experimental feature based on `claranet` team's request.

taoufik07 commented:
@Shr3ps I've resolved some conflicts, maybe you want to regenerate the README.

Shr3ps committed:
Update code to keep it retro-compatible with Terraform 1.2

Shr3ps committed:
Update for_each code (use name as key instead of index)
